### ISSUE TITLE
Merge diffeq - CORE

### ISSUE THREAD

**Issue by pnkraemer** [probabilistic-numerics/probnum#29]
Merge `randomvariable` class as well as other core functionalities (`quadrature`, `util`) together with existing one. 

* [ ] Mutual code review
* [x] Merging the tests
* [ ] Finalising the code merge

**Comment by JonathanWenger** [comment]
Here are some thoughts on where modules from `pn-diffeq` could go:
```
- diffeq
   - solver
   - statespace
- filtsmooth
    - bayesfilt
       - bayesianfilter.py
       - gaussianfilter
- linalg
- quad
- optim
   - bayesopt
   - classic/numerical
     - optimize.py (optimiser.py)
     - ....
- prob
   - covfuns
        - covariancefunctions.py (covariances.py)
        - brownianmotion.py
   - sampling
      - metropolishastings
- utils
```
Let me know what you think @pnkraemer !

**Comment by JonathanWenger** [comment]
For optimization I could also imagine a structure like
```
- optim
   - bayesopt
   - graddescent
   - newtontype
   - linesearch
   - optimize.py
```
Then `optimize.py` contains the interface/convenience functions and all the other folders contain the implementations. Similarly one could structure `quad` the same way:
```
- quad
   - integrate.py
   - bayesquad
   - clenshawcurtis
```

**Comment by pnkraemer** [comment]
I like that idea. I propose a few changes, roughly like this:
```
- diffeq
   - odesolver.py
   - ode.py
   - stepsize.py
   - odefilter/
      - prior.py
      - gaussianfilter.py
   - rungekutta/    (some basic implementations of standard runge kutta)
- filtsmooth
   - statespace
   - gaussfiltsmooth
       - kalman.py    (Kalman filter and smoother)
       - extendedkalman.py (Extended Kalman filter variants; filter and smoother)
       - ...
- linalg
- quad
   - quadrature.py   (some generic quadrature interface; nodes, weights, and a compute() function)
   - bayesquad/
      - bayesquad.py (interface plus some other respective modules)
   - polyn/
     - clenshawcurtis.py  ( plus whatever comes up)
- optim
   - bayesopt/ 
   - classic/optimization ( I would try to find a better name than "classic", something more descriptive)
     - optimize.py (optimiser.py)
     - ....
- prob
- sampling/
   - metropolishastings/
      - ...
- utils
   - covfuns
        - covariancefunctions.py (covariances.py)
        - brownianmotion.py
   - arrayutils (<- utils.py)
```

Essentially: `covfuns` is something auxiliary for me and not directly related to `prob/`.  `sampling` would make sense as its own module IMO, since it will likely outgrow `prob` very soon. `quad/`modules should share some common quadrature interface (nodes, weights). Filtering and smoothing in `filtsmooth` are algorithmically too similar to separate them strictly. The more reasonable separation here is Gaussian vs. Non-Gaussian.

**Comment by JonathanWenger** [comment]
New structure implemented in 6be4468.
